### DISCUSSION TITLE
Upfront generated token not working

### DISCUSSION THREAD

**Discussion by Regrau** [k3s-io/k3s#7156] [Q&A]
Hi everyone,

I'm installing a k3s server with the following settings.
```
    export K3S_KUBECONFIG_MODE="644"
    export K3S_CLUSTER_INIT="true"
    export K3S_AGENT_TOKEN_FILE=/usr/local/etc/k3s-agent-key
    export INSTALL_K3S_EXEC=" --disable servicelb --disable traefik"
    curl -sfL https://get.k3s.io | sh -
```

I have pre generated the k3s-agent-key via `k3s token generate`.

Now when I try to join a new agent to the cluster with the same key I get told that my node is unauthorized.
The error resolves when I explicitly create the token via `k3s token create $(cat /usr/local/etc/k3s-agent-key)` on the server. But it does not seem like I'm supposed to.

[From the documentation](https://docs.k3s.io/cli/server#:~:text=token%2Dfile%20value-,K3S_AGENT_TOKEN_FILE,-File%20containing%20the) I'd expect that the key will be usable after the server started. Am I misunderstanding something or do I handle the installation in a wrong way?

**Comment by brandond** [comment]
Check the docs: 
* https://docs.k3s.io/cli/token#k3s-token-generate
* https://docs.k3s.io/cli/token#k3s-token-create-token

Note that generate does not actually create the token, that's what create is for. If you don't pass a token to create, it will both generate and create a token for you.

**Reply by Regrau** [reply]
Hi, thanks for answering.

I did check the docs for the token subcommand. My question was more focused on the fact that while I'm creating a server I'm passing the `K3S_AGENT_TOKEN_FILE`.
It seemed to me that this will `create` the token in the cluster.

My goal here is to start the k3s-server with an already created agent token so I can start adding nodes without creating the secret manually.
I can script the token creation but it feels like a functionality that should be supported out of the box and I'm just missing the way to use it.

**Reply by brandond** [reply]
Ah I see, sorry. I thought you were referring to the new temporary tokens.

By default the agent token does not exist; it's just the same thing as the server token. If you want to use a distinct agent token, you must pass a value, or if you're specifying a path to a file, you must populate the file contents yourself.

**Reply by Regrau** [reply]
See, that's what I did.

Basically I do my deploy in this manner

```
    #This part I do manually and distribute the file later to my machines
    k3s token generate > /usr/local/etc/k3s-agent-key

    # This part is scripted
    export K3S_KUBECONFIG_MODE="644"
    export K3S_CLUSTER_INIT="true"
    export K3S_AGENT_TOKEN_FILE=/usr/local/etc/k3s-agent-key
    export INSTALL_K3S_EXEC=" --disable servicelb --disable traefik"
    curl -sfL https://get.k3s.io | sh -
    
    #On start up of a k3s agent machine I do the following
    export K3S_KUBECONFIG_MODE="644"
    export K3S_URL=https://$SERVER_HOST:6443
    export K3S_TOKEN_FILE=/usr/local/etc/k3s-agent-key
    curl -sfL https://get.k3s.io | sh -
```

An in such a way I'd expect the server to pick that file up and create a separate agent-token. But when the server is up I can't join agents
using the value inside the `/usr/local/etc/k3s-agent-key` file.

When I do the server declines nonchalantly with a 401 Unauthorized.
And after trying a while I'm still not sure if it's a bug or a OSI 8 Error

**Reply by brandond** [reply]
Can you check `/var/lib/rancher/k3s/server/cred/passwd` on the servers, and confirm that the password (the first field) for the `node` entry matches the token you put in `/usr/local/etc/k3s-agent-key` on the servers, and `/usr/local/etc/k3s-agent-key` on the agents?
